File: dsa-problems/LinkedLists/solutions/LNK-005-shuttle-route-alternating-reverse.py
import sys

class ListNode:
    def __init__(self, val=0):
        self.val = val
        self.next = None

def alternating_reverse(head: ListNode, l: int, k: int) -> ListNode:
    if not head or k <= 1:
        return head
    
    dummy = ListNode(0)
    dummy.next = head
    prev = dummy
    
    # Move to start position l
    for _ in range(l - 1):
        if not prev.next:
            return head
        prev = prev.next
        
    reverse_turn = True
    
    while prev.next:
        if reverse_turn:
            # Reverse next k nodes
            tail = prev.next
            curr = tail.next
            count = 1
            while curr and count < k:
                temp = curr.next
                curr.next = prev.next
                prev.next = curr
                tail.next = temp
                curr = temp
                count += 1
            prev = tail # Move prev to end of reversed block
        else:
            # Skip k nodes
            count = 0
            while prev.next and count < k:
                prev = prev.next
                count += 1
        
        reverse_turn = not reverse_turn
        
    return dummy.next

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    iterator = iter(data)
    try:
        n = int(next(iterator))
        dummy = ListNode()
        cur = dummy
        for _ in range(n):
            cur.next = ListNode(int(next(iterator)))
            cur = cur.next
            
        l = int(next(iterator))
        k = int(next(iterator))
        
        head = alternating_reverse(dummy.next, l, k)
        
        out = []
        while head:
            out.append(str(head.val))
            head = head.next
        print(" ".join(out))
    except StopIteration:
        pass

if __name__ == "__main__":
    main()

File: dsa-problems/LinkedLists/solutions/javascript/LNK-005-shuttle-route-alternating-reverse.js
const readline = require("readline");

class ListNode {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

function alternatingReverse(head, l, k) {
  if (!head || k <= 1) return head;

  const dummy = new ListNode(0);
  dummy.next = head;
  let prev = dummy;

  // Move to l-1
  for (let i = 0; i < l - 1; i++) {
    if (!prev.next) return head;
    prev = prev.next;
  }

  let reverse = true;
  while (prev.next) {
    if (reverse) {
      let tail = prev.next;
      let curr = tail.next;
      let count = 1;
      while (curr && count < k) {
        let temp = curr.next;
        curr.next = prev.next;
        prev.next = curr;
        tail.next = temp;
        curr = temp;
        count++;
      }
      prev = tail;
    } else {
      let count = 0;
      while (prev.next && count < k) {
        prev = prev.next;
        count++;
      }
    }
    reverse = !reverse;
  }
  return dummy.next;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/).filter(x => x)));
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  
  const dummy = new ListNode(0);
  let cur = dummy;
  for (let i = 0; i < n; i++) {
    cur.next = new ListNode(parseInt(data[idx++], 10));
    cur = cur.next;
  }
  
  if (idx < data.length) {
      const l = parseInt(data[idx++], 10);
      const k = parseInt(data[idx++], 10);

      let head = alternatingReverse(dummy.next, l, k);
      const out = [];
      while (head) {
        out.push(head.val);
        head = head.next;
      }
      console.log(out.join(" "));
  }
});
